src/main/java/com/conveyal/r5/transit/TransferFinder.java

```diff
@@ -181,33 +179,71 @@ public void findTransfers () {
+     * transfer targets. Two details help limit this possibility.
+     *
+     * First, the filtering should be called separately for each source stop, which helps address concerns about
+     * certain irregular route shapes (such as a U-shaped or looping routes where two different target stops might be
+     * optimal depending on the direction of travel). The filtering will be correct unless all stops on the source
+     * pattern are closer to one of the target stops than the other. Examples:
+     * 1. Depending on the destination, transfers from the WMATA Q6 to the same pattern on the U-shaped Red Line could
+     * be logical at Rockville or Wheaton. But these are far apart, with multiple different bus stops available as
+     * transfer sources for each rail station, so results should be correct.
+     * 2. Transfers from Singapore bus routes to the Downtown Line at Rochor (DT13) or Jalan Besar (DT22). If someone
+     * wants to transfer at Jalan Besar and travel onward to Bedok North (DT29), it is conceivable this method could
+     * erroneously filter out their preferred transfer, forcing them to ride around the Downtown loop. But all the
+     * bus's stops would need to be closer to stops DT1-13 than DT22-29, which seems unlikely given the Singapore
+     * street network.
+     *
+     * Second, when considering transfers from a given source stop, the filtering retains a different nearby stop as
+     * a potential target. If the source stop is also being considered as a target (i.e. it is served by both the
+     * source and target patterns), it will generally be the closest (except conceivably in edge cases related to
+     * street linking) and is retained separately. This logic helps handle situations with routes that provide
+     * service in different "directions" on the same pattern -- for example, a journey from Paulina (Brown) to Howard
+     * (Purple) on the CTA El. Assume the CTA codes Purple Line Express service as a single pattern (south, around
+     * the Loop, and back north), with separate stops for the northbound and southbound boarding platforms at Belmont.
+     * Transferring from the Brown Line southbound at Belmont, the closest stop on the Purple Line Express pattern would
+     * be the same platform. But for this journey, the desired transfer would actually imply switching platforms at
+     * Belmont. Ignoring the source transfer in the search for the nearest stop, then adding it back at the end,
+     * should handle this case.
+     *
+     * So, this filtering can be incorrect in certain rare situations, but anecdotally it speeds up computation by up
+     * to 40 percent. TODO confirm this performance gain with more systematic measurement, and look into other ways to
+     * optimize transfers (or why the transfers make routing so slow).
+     *
+     * @param timesToReachedStops map from target stop index to distance (along the street network) to it
```

Comment:
We should call attention to the fact that this method has side effects on one of its parameters, which is usually assumed not to be the case.

Reply:
https://github.com/conveyal/r5/issues/619